### PR TITLE
Fix: Always ignore defaults unless explicitly passed (fixes #5547)

### LINKED ISSUE #5547: `--no-ignore` should not un-ignore node_modules
**What version of ESLint are you using?**
Master

**What configuration and parser (Espree, Babel-ESLint, etc.) are you using?**
Espree

**What did you do? Please include the actual source code causing the issue.**
Ran eslint with `--no-ignore` flag

**What did you expect to happen?**
Ignore settings from `.eslintignore` to be disregarded, as indicated in docs.

**What actually happened? Please include the actual, raw output from ESLint.**
ESLint linted all files within `node_modules`.

Despite having [tests](https://github.com/eslint/eslint/blob/d727f17f4efc888ed0e59259a4e4cac01a42f0d9/tests/lib/ignored-paths.js#L375-L383) to assert that `--no-ignore` will cause `node_modules` to be linted, I think this is not the behavior we expect or want.  

The --no-ignore docs say:
> --no-ignore
> Disables excluding of files from .eslintignore and --ignore-path files.

They don't say anything about also disabling default ignores.

After https://github.com/eslint/eslint/issues/5410 is fixed, it will be possible to lint within node_modules by using a negation pattern.  (e.g. `!/node_modules/package-to-lint/`), at which time I think it would make sense to keep `--no-ignore` from linting `/node_modules/` and `/bower_components/`.

Does the team agree?

### PR DESCRIPTION
**Pull request by IanVS**
I was getting some strange, seemingly unrelated failures on my machine (even on master) about valid-jsdoc.  Hopefully it's just something wonky on my box.

### CODE REVIEW

**Comment by IanVS** [comment]
Since Travis seems to be passing, this is the error I am getting (just recording here so I can investigate deeper at some later time):

```
  10386 passing (1m)
  2 failing

  1) valid-jsdoc valid /**
 * Foo
 * @module module-name
 */
function foo() {}:

      AssertionError: Should have no errors but had 1: [ { ruleId: 'valid-jsdoc',
    severity: 1,
    message: 'JSDoc syntax error.',
    line: 1,
    column: 1,
    nodeType: 'Block',
    source: '/**' } ]
      + expected - actual

      -1
      +0

      at testValidTemplate (lib/testers/rule-tester.js:9:9447)
      at Context.<anonymous> (lib/testers/rule-tester.js:9:14196)

  2) valid-jsdoc valid /**
 * Foo
 * @alias module:module-name
 */
function foo() {}:

      AssertionError: Should have no errors but had 1: [ { ruleId: 'valid-jsdoc',
    severity: 1,
    message: 'JSDoc syntax error.',
    line: 1,
    column: 1,
    nodeType: 'Block',
    source: '/**' } ]
      + expected - actual

      -1
      +0

      at testValidTemplate (lib/testers/rule-tester.js:9:9447)
      at Context.<anonymous> (lib/testers/rule-tester.js:9:14196)
```

**Comment by alberto** [comment]
Thanks @IanVS. I was in the process of putting a PR together to discuss with you about this, but didn't find the time.

I think right now `ig.default` contains all ignored patterns and `ig.custom` just the custom ones from `.eslintignore`. Do you think it makes sense to change it so that `default` contains just our defaults and `custom` (or `all` if you will) contains everything? So, we would use one or the other depending on whether we use `--no-ignore`

**Comment by IanVS** [comment]
Hm, that's an interesting thought, and it might help improve the clarity of the code.  I guess my suggestion would be to merge this if it looks okay to the reviewers, and then we can use it and its tests as a baseline to refactor as you suggest.

**Comment by platinumazure** [comment]
I like @alberto's suggestion and @IanVS's implementation plan. Anything else holding this back?

**Comment by IanVS** [comment]
Just needs a review from a reviewer.

Hm, there's no @ mention for "reviewers-team".

File: tests/lib/util/glob-util.js [hunk 1]
```diff
@@ -221,6 +221,50 @@ describe("globUtil", function() {
...
+            var directory = getFixturePath("glob-util", "hidden");
+            var patterns = [directory];
+            var result = globUtil.listFilesToProcess(patterns, {
+                cwd: getFixturePath()
+            });
+
+            assert.equal(result.length, 0);
+        });
+
+        it("should ignore and warn for default ignored files when passed explicitly", function() {
```
Comment (alberto) [inline tests/lib/util/glob-util.js]: @nzakas Is this the expected behavior? I understood otherwise from these [comments](https://github.com/eslint/eslint/issues/4828#issuecomment-194535273)

File: tests/lib/cli-engine.js [hunk 2]
```diff
@@ -254,6 +254,21 @@ describe("CLIEngine", function() {
             });
         });
 
+        // https://github.com/eslint/eslint/issues/5547
+        it("should respect default ignore rules, even with --no-ignore", function() {
```
Comment (alberto) [inline tests/lib/cli-engine.js]: Same question as below here regarding explicitly passed files.

File: tests/lib/cli-engine.js [hunk 3]
```diff
@@ -254,6 +254,21 @@ describe("CLIEngine", function() {
...
+            engine = new CLIEngine({
+                cwd: getFixturePath(),
+                ignore: false
+            });
+
+            var report = engine.executeOnText("var bar = foo;", "node_modules/passing.js");
+
+            assert.equal(report.results.length, 1);
+            assert.equal(report.results[0].filePath, getFixturePath("node_modules/passing.js"));
+            assert.equal(report.results[0].messages[0].message, "File ignored because of a matching ignore pattern. Use --no-ignore to override.");
```
Comment (alberto) [inline tests/lib/cli-engine.js]: This warning proposes an invalid solution for default files, you are already using --no-ignore.

File: tests/lib/util/glob-util.js (hunk 1 above)
Comment (nzakas) [inline tests/lib/util/glob-util.js]: I think that's the big we are fixing here, no?

Comment (IanVS) [inline tests/lib/util/glob-util.js]: I think we may be getting mixed up on a few related issues.  This PR is to fix https://github.com/eslint/eslint/issues/5547, which is to say, to keep ignoring default-ignored files even if the `--no-ignore` flag is passed.  

The current behavior has always been to print a warning if an explicitly passed file is ignored due to an ignore rule.  https://github.com/eslint/eslint/issues/4828#issuecomment-194535273 is a separate issue and will need greater structural changes to accomplish.

File: tests/lib/cli-engine.js (hunk 3 above)
Comment (IanVS) [inline tests/lib/cli-engine.js]: Good point, we should have two separate messages for default-ignored and regular files.  Although, https://github.com/eslint/eslint/issues/4828#issuecomment-194535273 may make all of this "warning about ignored files" stuff obsolete, since it looks like we will no longer do that.

File: tests/lib/util/glob-util.js (hunk 1 above)
Comment (nzakas) [inline tests/lib/util/glob-util.js]: :+1:

Comment (alberto) [inline tests/lib/util/glob-util.js]: I'm not sure we should add a test for a behaviour we think is a bug, but yeah, it's a different issue.

**Comment by nzakas** [comment]
@IanVS are you still working on this?

**Comment by IanVS** [comment]
I'm unsure what the required tasks are, to be honest.  Do I need to create a different warning message in cases where it's necessary to use a special override pattern to unignore a default ignore?  It seems like it may be short lived if we stop using such warnings at all, as seems to be the direction we've decided to go.

**Comment by alberto** [comment]
@IanVS I'm not sure when that would be done (we don't even have an specific issue for it and I'm not 100% sure it's without problems), so I'd say it's better to add it for now.

**Comment by nzakas** [comment]
@alberto is this ready to merge in your opinion?

I've gotten a bit lost here, so don't want to assume anything. :)

**Comment by IanVS** [comment]
I need to make a small update to the warning messages.  Working on it now.

**Comment by IanVS** [comment]
I just now signed the jQuery CLA.  

@alberto once you give the 👍 to my latest commit, I'll squash commits.  Wanted to make it easier to see the new changes.

File: lib/cli-engine.js [hunk 4]
```diff
@@ -300,18 +300,31 @@ function processFile(filename, configHelper, options) {
...
+ * @param {string} baseDir  - Absolute path of base directory
+ * @returns {Result}           Result with single warning
  * @private
  */
-function createIgnoreResult(filePath) {
+function createIgnoreResult(filePath, baseDir) {
+    var message;
+
+    if (/^\./.test(path.basename(filePath))) {
+        message = "Hidden file ignored by default.  Use '--ignore-pattern !.*' to override.";
```
Comment (alberto) [inline lib/cli-engine.js]: I am not sure we should use relative paths here, since we use absolute paths when reporting.

I think it would be better not to recommend to uningnore everything, but just the current file. Incidentally, you could then use a single message for all 3 cases: "File ignored by default. Use '--ignore-pattern \"!" + filePath +"\"' to overrride." (I also added necessary quotes around the pattern)

And maybe use `ignoredPaths.contains(filePath, "default")` instead of explicitly checking each ignore pattern?

Comment (alberto) [inline lib/cli-engine.js]: Forget what I said. Just realized ignore patterns have to be relative. :)

Comment (IanVS) [inline lib/cli-engine.js]: I was trying to keep this as simple as possible, knowing that the fancier I make it the more likely there will be bugs.  It's already pretty ugly, and hopefully can be cleaned up when we refactor to simplify default ignores.  For now, would you be opposed to a simpler message like:

```
File ignored by default.  Use a negated --ignore-pattern (like '!<filename>') to override."
```

Comment (alberto) [inline lib/cli-engine.js]: That's what I was proposing initially. Unfortunately it won't work for files in nested folders in `node_modules` or `bower_components`. You can't just unignore `node_modules/foo/bar.js`, you have to unignore `node_modules/foo/` :/

Comment (IanVS) [inline lib/cli-engine.js]: I think you misunderstand.  I mean that message literally.  Not using a variable.  Let the user figure out what the pattern should be, after giving them the hint to use a `!`.

Comment (alberto) [inline lib/cli-engine.js]: To be clear, I'm proposing to adapt it like this:

```
var relativePath = path.relative(baseDir, filePath));

if (/^\./.test(path.basename(filePath))) {
  message = "File ignored by default.  Use '--ignore-pattern \"!" + relativePath + "\"' to override.";
} else if (baseDir && /^node_modules/.test(relativePath) {
  message = "File ignored by default. Use '--ignore-pattern \"!node_modules/*\"' to override.";
} else if (baseDir && /^bower_components/.test(path.relative(baseDir, filePath))) {
  message = "File ignored by default. Use '--ignore-pattern \"!bower_components/*\"' to override.";
} else {
  message = "File ignored because of a matching ignore pattern. Use '--no-ignore' to override.";
}
```

Comment (alberto) [inline lib/cli-engine.js]: I did, but to be able to figure out the right pattern yourself (as I said, you can't simply unignore the file and I think that's what people would try given that message), you'd have to know we are using `node_modules/*` to ignore.

Comment (IanVS) [inline lib/cli-engine.js]: Yes, but I would like to avoid using `relativePath` directly in the message because baseDir may not always be provided and there may be other edge cases and conditions that cause trouble.  I'd rather replace all of that logic with:

```
var isHidden = /^\./.test(path.basename(filePath));
var isInNodeModules = baseDir && /^node_modules/.test(path.relative(baseDir, filePath));
var isInBowerComponents = baseDir && /^bower_components/.test(path.relative(baseDir, filePath));

if (isHidden || isInNodeModules || isInBowerComponents) {
  message = "File ignored by default.  Use a negated --ignore-pattern (like '!<relative/path/to/filename>') to override."";
} else {
  message = "File ignored because of a matching ignore pattern. Use '--no-ignore' to override.";
}
```

This is still more complicated than I'd like and still relies on `baseDir` for `node_modules` and `bower_components` (no way around that), but does not require it for hidden files.  The more explicit we are in the error message (by giving an exact flag to use), the greater chances for confusion if/when that suggestion does not work.

Comment (IanVS) [inline lib/cli-engine.js]: Are you sure `!node_modules/foo/bar.js` won't work?  That would be annoying. . .

**Comment by IanVS** [comment]
Hm, just found a problem that nested files inside `node_modules` (like `node_modules/foo/bar.js`) are linted when specified directly, instead of being ignored. Will need to work on this a bit more I guess. :(

File: lib/cli-engine.js (hunk 4 above)
Comment (alberto) [inline lib/cli-engine.js]: You can try it on your branch with:

```
eslint node_modules/escope/gulpfile.js --ignore-pattern '!node_modules/escope/gulpfile.js'
```

vs:
```
eslint node_modules/escope/gulpfile.js --ignore-pattern '!node_modules/*'
```

Also, bash needs single, not double quotes with "!" for some reason. :(

**Comment by alberto** [comment]
@IanVS If we change our default ignore patterns to:

```
"/node_modules/**/*.js",
"/bower_components/**/*.js"
```

This should allow directly unignoring individual files. :)

**Comment by IanVS** [comment]
Yes, but I would be worried about other unintended consequences of doing that.  For example if someone uses `--ext .es6` and there happen to be `.es6` files in a node module.

I think that leaves us with the following as the best-case we can hope for at this time.  Do you agree?

```js
var isHidden = /^\./.test(path.basename(filePath));
var isInNodeModules = baseDir && /^node_modules/.test(path.relative(baseDir, filePath));
var isInBowerComponents = baseDir && /^bower_components/.test(path.relative(baseDir, filePath));

if (isHidden) {
  message = "File ignored by default.  Use a negated --ignore-pattern (like '!<relative/path/to/filename>') to override.";
} else if (isInNodeModules) {
  message = "File ignored by default. Use '--ignore-pattern \"!node_modules/*\"' to override.";
} else if (isInBowerComponents) {
  message = "File ignored by default. Use '--ignore-pattern \"!bower_components/*\"' to override.";
} else {
  message = "File ignored because of a matching ignore pattern. Use '--no-ignore' to override.";
}
```

**Comment by alberto** [comment]
> [...] I would like to avoid using relativePath directly in the message because baseDir may not always be provided and there may be other edge cases and conditions that cause trouble.

I would expect this would work the same as `ignored-paths`, which uses `process.cwd()` by default, since this is directly related. But if you still think it's better not to do it, it's not a big deal.

> Yes, but I would be worried about other unintended consequences of doing that. For example if someone uses --ext .es6 and there happen to be .es6 files in a node module.

Ok, let's leave that out of this issue.

> Do you agree?

Sorry, I was wrong. You need to use single quotes instead of double quotes around the `!node_modules/*` and `!bower_components/*`patterns, otherwise bash tries to expand the `!`

**Comment by IanVS** [comment]
OK @alberto, I've updated the wording as we've discussed.  Anything else you see that I should take care of?  Thanks for working through this together.

**Comment by nzakas** [comment]
Since we are doing a release tomorrow and it looks like there might be some outstanding ignore issues (#5979), let's hold off on merging this at least until Tuesday (to be clear of any latch release), and then we can have two weeks of baking on master to be sure of the changes.

**Comment by alberto** [comment]
LGTM. 👍 

Thanks @IanVS for keeping up with this.

**Comment by kaelzhang** [comment]
Sorry for the ignore issues I brought it to you guys, and for my delay of fixing these issues, that there is a time lag between us, and I could only check issues and make the fixing while you are asleep.

However, all confirmed bugs of `node-ignore` will always be fixed within 24 hours. You have my word.

**Comment by alberto** [comment]
Thanks @kaelzhang, no need to apologize for bugs, they are unavoidable. 

And we are also spread over the world, have our jobs, time of sleep, have our own lives and all, so we don't expect a SLA from anyone. :)

Thanks for your work. ❤️
